components/ILIAS/OpenIdConnect/classes/class.ilOpenIdConnectSettingsGUI.php

```diff
@@ -498,68 +438,231 @@ protected function prepareRoleSelection(bool $a_with_select_option = true): arra
-            $text_form->setValue($this->settings->getProfileMappingFieldValue($field));
-            $form->addItem($text_form);
+    protected function saveScopes(): void
+    {
+        $this->checkAccess('write');
+        $validation = false;
+        $type = null;
+        $url = null;
+        $custom_scopes = [];
+
+        $form = $this->initScopesForm();
+        if ($this->request->getMethod() == "POST") {
+            $request_form = $form->withRequest($this->request);
+            $result = $request_form->getData();
+            if (is_null($result)) {
+                $this->mainTemplate->setOnScreenMessage('failure', $this->lng->txt('err_check_input'));
+                $this->scopes();
+                return;
+            } else {
+                foreach ($form->getInputs() as $group => $groups) {
+                    foreach ($groups->getInputs() as $key => $input) {
+                        $dedicated_name = $input->getDedicatedName();
+                        $result_data = $result[$group][$key];
+                        if($dedicated_name === 'validate_scopes') {
+                            $type = (int) $result_data[0];
+                            $url = array_pop($result_data[1]);
+                        } elseif($dedicated_name === 'custom_scope') {
+                            $custom_scopes = $result_data;
+                        }
+                    }
+                }
+            }
+            if($url === null && $type === ilOpenIdConnectSettings::URL_VALIDATION_PROVIDER) {
+                $url = $this->settings->getProvider();
+            }
+            $validation = $this->validateDiscoveryUrl($type, $url, $custom_scopes);
+        }
 
-            $checkbox_form = new ilCheckboxInputGUI('');
-            $checkbox_form->setValue("1");
-            $checkbox_form->setPostVar($field . "_update");
-            $checkbox_form->setChecked($this->settings->getProfileMappingFieldUpdate($field));
-            $checkbox_form->setOptionTitle($this->lng->txt('auth_oidc_update_field_info'));
-            $form->addItem($checkbox_form);
+        if ($validation) {
+            $this->settings->setAdditionalScopes((array) $custom_scopes);
+            $this->settings->setValidateScopes((int) $type);
+            if (ilOpenIdConnectSettings::URL_VALIDATION_CUSTOM === $this->settings->getValidateScopes()) {
+                $this->settings->setCustomDiscoveryUrl($url);
+            }
+            $this->settings->save();
+            $this->mainTemplate->setOnScreenMessage('success', $this->lng->txt('settings_saved'), true);
+            $this->ctrl->redirect($this, 'scopes');
         }
 
-        if ($this->checkAccessBool('write')) {
-            $form->addCommandButton('saveProfile', $this->lng->txt('save'));
+        if(strlen($this->failed_validation_messages) > 0) {
+            $this->failed_validation_messages = $this->lng->txt('err_check_input') . '<br/>' . $this->failed_validation_messages;
```

Comment:
Maybe it is possible to avoid this `<br />` here.

Reply:
I would be open for suggestions.

Reply:
I suggest appending the lang_keys to the redirect url (comma separated, with ilCtrl::setParameter[byClass]) and the use multiple "UI > Messagebox > Failure" to present the exploded language variables.